File: x_ide/src/org/alice/ide/cascade/fillerinners/BooleanFillerInner.java
/*
 * Copyright (c) 2006-2009, Carnegie Mellon University. All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 * 
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 
 * 3. Products derived from the software may not be called "Alice",
 *    nor may "Alice" appear in their name, without prior written
 *    permission of Carnegie Mellon University.
 * 
 * 4. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *    "This product includes software developed by Carnegie Mellon University"
 */
package org.alice.ide.cascade.fillerinners;

/**
 * @author Dennis Cosgrove
 */
public class BooleanFillerInner extends ExpressionFillerInner {
	public BooleanFillerInner() {
		super( edu.cmu.cs.dennisc.alice.ast.TypeDeclaredInJava.BOOLEAN_OBJECT_TYPE, edu.cmu.cs.dennisc.alice.ast.BooleanLiteral.class );
	}
	@Override
	public void addFillIns( edu.cmu.cs.dennisc.cascade.Blank blank ) {
		final edu.cmu.cs.dennisc.alice.ast.ConditionalInfixExpression.Operator[] CONDITIONAL_OPERATORS = {
				edu.cmu.cs.dennisc.alice.ast.ConditionalInfixExpression.Operator.AND, 	
				edu.cmu.cs.dennisc.alice.ast.ConditionalInfixExpression.Operator.OR, 	
		};
		//todo: relational
		edu.cmu.cs.dennisc.alice.ast.Expression previousExpression = org.alice.ide.IDE.getSingleton().getPreviousExpression();
		final boolean isTop = blank.getParentFillIn() == null;
		if( isTop ) {
			if( previousExpression instanceof edu.cmu.cs.dennisc.alice.ast.ConditionalInfixExpression ) {
				edu.cmu.cs.dennisc.alice.ast.ConditionalInfixExpression previousConditionalInfixExpression = (edu.cmu.cs.dennisc.alice.ast.ConditionalInfixExpression)previousExpression;
				edu.cmu.cs.dennisc.alice.ast.Expression leftOperand = previousConditionalInfixExpression.leftOperand.getValue();
				edu.cmu.cs.dennisc.alice.ast.ConditionalInfixExpression.Operator prevOperator = previousConditionalInfixExpression.operator.getValue();
				edu.cmu.cs.dennisc.alice.ast.Expression rightOperand = previousConditionalInfixExpression.rightOperand.getValue();
				for( edu.cmu.cs.dennisc.alice.ast.ConditionalInfixExpression.Operator operator : CONDITIONAL_OPERATORS ) {
					if( operator != prevOperator ) {
						blank.addFillIn( new org.alice.ide.cascade.LabeledExpressionFillIn( new edu.cmu.cs.dennisc.alice.ast.ConditionalInfixExpression( leftOperand, operator, rightOperand ), "(replace operator)" ) );
					}
				}
				blank.addSeparator();
				blank.addFillIn( new org.alice.ide.cascade.LabeledExpressionFillIn( leftOperand, "(reduce to left operand only)" ) );
				blank.addFillIn( new org.alice.ide.cascade.LabeledExpressionFillIn( rightOperand, "(reduce to right operand only)" ) );
				blank.addSeparator();
			} else if( previousExpression instanceof edu.cmu.cs.dennisc.alice.ast.LogicalComplement ) {
				edu.cmu.cs.dennisc.alice.ast.LogicalComplement previousLogicalComplement = (edu.cmu.cs.dennisc.alice.ast.LogicalComplement)previousExpression;
				blank.addFillIn( new org.alice.ide.cascade.LabeledExpressionFillIn( previousLogicalComplement.operand.getValue(), "(reduce to inner operand only)" ) );
				blank.addSeparator();
			}
		}
		this.addExpressionFillIn( blank, true );
		this.addExpressionFillIn( blank, false );
		blank.addSeparator();
		if( isTop ) {
	//			blank.addFillIn( new edu.cmu.cs.dennisc.cascade.MenuFillIn( "Random" ) {
	//			@Override
	//			protected void addChildrenToBlank(edu.cmu.cs.dennisc.cascade.Blank blank) {
	//				addNodeChildForMethod( blank, RANDOM_UTILITIES_TYPE_EXPRESSION, "nextBoolean" );
	//			}
	//		} );
			addNodeChildForMethod( blank, RANDOM_UTILITIES_TYPE_EXPRESSION, "nextBoolean" );
			
			blank.addSeparator();
			if( previousExpression != null ) {
				blank.addFillIn( new org.alice.ide.cascade.SimpleExpressionFillIn< edu.cmu.cs.dennisc.alice.ast.LogicalComplement >( new edu.cmu.cs.dennisc.alice.ast.LogicalComplement( previousExpression ) ) );
			}
			blank.addFillIn( new org.alice.ide.cascade.IncompleteLogicalComplementFillIn() );
			blank.addSeparator();
			if( blank.getParentFillIn() == null ) {
				if( previousExpression != null ) {
					for( edu.cmu.cs.dennisc.alice.ast.ConditionalInfixExpression.Operator operator : CONDITIONAL_OPERATORS ) {
						blank.addFillIn( new org.alice.ide.cascade.MostlyDeterminedConditionalInfixExpressionFillIn( previousExpression, operator ) );
					}
				}
			}
			for( edu.cmu.cs.dennisc.alice.ast.ConditionalInfixExpression.Operator operator : edu.cmu.cs.dennisc.alice.ast.ConditionalInfixExpression.Operator.values() ) {
				blank.addFillIn( new org.alice.ide.cascade.IncompleteConditionalExpressionFillIn( operator ) );
			}
			blank.addSeparator();
			blank.addFillIn( new edu.cmu.cs.dennisc.cascade.MenuFillIn( "Relational (Real Number) { ==, !=, <, <=, >=, > }" ) {
				@Override
				protected void addChildrenToBlank(edu.cmu.cs.dennisc.cascade.Blank blank) {
					for( edu.cmu.cs.dennisc.alice.ast.RelationalInfixExpression.Operator operator : edu.cmu.cs.dennisc.alice.ast.RelationalInfixExpression.Operator.values() ) {
						blank.addFillIn( new org.alice.ide.cascade.IncompleteRelationalExpressionFillIn( Number.class, operator ) );
					}
				}
			} );
			blank.addFillIn( new edu.cmu.cs.dennisc.cascade.MenuFillIn( "Relational (Integer) { ==, !=, <, <=, >=, > }" ) {
				@Override
				protected void addChildrenToBlank(edu.cmu.cs.dennisc.cascade.Blank blank) {
					for( edu.cmu.cs.dennisc.alice.ast.RelationalInfixExpression.Operator operator : edu.cmu.cs.dennisc.alice.ast.RelationalInfixExpression.Operator.values() ) {
						blank.addFillIn( new org.alice.ide.cascade.IncompleteRelationalExpressionFillIn( Integer.class, operator ) );
					}
				}
			} );
			blank.addSeparator();
		}
	}
}

File: ide_py_foundation/src/ecc/dennisc/alice/ide/cascade/fillerinners/BooleanFillerInner.py
#print "-->", __name__

from ExpressionFillerInner import *

class BooleanFillerInner( ExpressionFillerInner ):
	def __init__( self ):
		ExpressionFillerInner.__init__( self, ecc.dennisc.alice.ast.getType( java.lang.Boolean ), alice.ast.BooleanLiteral )
	def addFillIns( self, blank ):
		self._addExpressionFillIn( blank, True )
		self._addExpressionFillIn( blank, False )
		blank.addSeparator()
		blank.addChild( ecc.dennisc.alice.ide.cascade.ConditionalInfixExpressionFillIn() )
		blank.addChild( ecc.dennisc.alice.ide.cascade.RelationalInfixExpressionFillIn( "relational expressions { ==, !=, <, <=, >=, > } (Real Number)", java.lang.Double ) )
		blank.addChild( ecc.dennisc.alice.ide.cascade.RelationalInfixExpressionFillIn( "relational expressions { ==, !=, <, <=, >=, > } (Integer)", java.lang.Integer ) )


#class ConditionalMenuFillIn( edu.cmu.cs.dennisc.cascade.MenuFillIn ):
#	def __init__(self):
#		edu.cmu.cs.dennisc.cascade.MenuFillIn.__init__( self, "conditional { and, or }" )
#	def addChildrenToBlank(self, blank):
#		blank.addChild( self._createConditionalInfixExpressionFillIn( alice.ast.ConditionalInfixExpression.Operator.AND ) )
#		blank.addChild( self._createConditionalInfixExpressionFillIn( alice.ast.ConditionalInfixExpression.Operator.OR ) )
#	def _createConditionalInfixExpressionFillIn( self, operator ):
#		type = ecc.dennisc.alice.ast.getType( java.lang.Boolean )
#		return ecc.dennisc.alice.ide.cascade.InfixExpressionFillIn( alice.ast.ConditionalInfixExpression( alice.ide.editors.code.EmptyExpression( type ), operator, alice.ide.editors.code.EmptyExpression( type ) ) )


#class InfixExpressionFillIn2( edu.cmu.cs.dennisc.cascade.SimpleFillIn ):
#	def __init__( self, infixExpression, cls ):
#		edu.cmu.cs.dennisc.cascade.SimpleFillIn.__init__( self, infixExpression )
#		self._type = ecc.dennisc.alice.ast.getType( cls )
#
#	def createMenuProxy(self):
#		return alice.ide.editors.code.ExpressionPane( self.getModel() )

#class ConditionalMenuFillIn( edu.cmu.cs.dennisc.cascade.FillIn ):
#	def __init__(self):
#		edu.cmu.cs.dennisc.cascade.FillIn.__init__( self )
#		self._expression = alice.ast.ConditionalInfixExpression()
#	def addChildren(self):
#		self.addChild( ecc.dennisc.alice.ide.cascade.ExpressionPropertyBlank( self._expression.leftOperand ) )
#		self.addChild( OperatorBlank() )
#		self.addChild( ecc.dennisc.alice.ide.cascade.ExpressionPropertyBlank( self._expression.rightOperand ) )
#	def getValue(self):
#		rv = alice.ast.ConditionalInfixExpression()
#		rv.leftOperand.setValue( self.getChildren()[ 0 ].getSelectedFillIn().getValue() )
#		rv.operator.setValue( self.getChildren()[ 1 ].getSelectedFillIn().getValue().operator.getValue() )
#		rv.rightOperand.setValue( self.getChildren()[ 2 ].getSelectedFillIn().getValue() )
#		return rv
#	def createMenuProxy(self):
#		return javax.swing.JLabel( "conditional { and, or }" )

#print "<--", __name__
